### PR TITLE
Convert to typescript

### PR DESCRIPTION
**Pull request by justjake**
* Now typescript
* Use class, etc.

### CODE REVIEW

File: src/memjs/header.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    dataType: headerBuf.readUInt8(5),
+    status: headerBuf.readUInt16BE(6) as ResponseStatus, // TODO: wrong type?
+    totalBodyLength: headerBuf.readUInt32BE(8),
+    opaque: headerBuf.readUInt32BE(12),
+    cas: headerBuf.slice(16, 24),
+  };
+}
+
+/** toBuffer converts a JS memcache header object to a binary memcache header */
+export function toBuffer(header: Header) {
```
Comment (blackmad) [inline src/memjs/header.ts]:
```suggestion
export function toBuffer(header: Header): Buffer {
```

File: src/memjs/memjs.ts [hunk 2]
```diff
@@ -0,0 +1,1777 @@
+// # MemJS Memcache Client
```
Comment (blackmad) [inline src/memjs/memjs.ts]:
```suggestion
// MemTS Memcache Client
```

Comment (blackmad) [reply src/memjs/memjs.ts]: ;-)

File: src/memjs/memjs.ts [hunk 3]
```diff
@@ -0,0 +1,1777 @@
...
+  makeAmountInitialAndExpiration,
+  hashCode,
+  MaybeBuffer,
+  Message,
+} from "./utils";
+import * as constants from "./constants";
+import * as Utils from "./utils";
+import * as Header from "./header";
+
+function defaultKeyToServerHashFunction(servers: string[], key: string) {
```
Comment (blackmad) [inline src/memjs/memjs.ts]:
```suggestion
function defaultKeyToServerHashFunction(servers: string[], key: string): string {
```

File: src/memjs/memjs.ts [hunk 4]
```diff
@@ -0,0 +1,1777 @@
...
+// converts a call into a promise-returning one
+function promisify<Result>(
+  command: (callback: (error: Error | null, result: Result) => void) => void
+): Promise<Result> {
+  return new Promise(function (resolve, reject) {
+    command(function (err, result) {
+      err ? reject(err) : resolve(result);
+    });
+  });
+}
```
Comment (blackmad) [inline src/memjs/memjs.ts:44]: this could be native node promisify no?

File: src/memjs/memjs.ts [hunk 5]
```diff
@@ -0,0 +1,1777 @@
...
+  expires: number;
+  logger: { log: typeof console.log };
+  keyToServerHashFunction: typeof defaultKeyToServerHashFunction;
+}
+
+interface SerializerProp<Value, Extras> {
+  serializer: Serializer<Value, Extras>;
+}
+
+type IfBuffer<Value, Extras, IsBuffer, NotBuffer> = Value extends Buffer
```
Comment (blackmad) [inline src/memjs/memjs.ts:63]: comment?

Comment (blackmad) [reply src/memjs/memjs.ts:63]: can't quite parse this on first read

File: src/memjs/memjs.ts [hunk 6]
```diff
@@ -0,0 +1,1777 @@
...
+      logger: console,
+      keyToServerHashFunction: defaultKeyToServerHashFunction,
+    });
+
+    this.serializer = this.options.serializer || (noopSerializer as any);
+
+    // Store a mapping from hostport -> server so we can quickly get a server object from the serverKey returned by the hashing function
+    var serverMap: { [hostport: string]: Server } = {};
+    this.servers.forEach(function (server) {
+      serverMap[server.hostportString()] = server;
```
Comment (blackmad) [inline src/memjs/memjs.ts]: can't remember why I had a local var here, make this operate on this.serverMap directly?

File: src/memjs/memjs.ts [hunk 7]
```diff
@@ -0,0 +1,1777 @@
...
+  serverKeyToServer(serverKey: string) {
+    return this.serverMap[serverKey];
+  }
+
+  /**
+   * Given a key to look up in memcache, return a serverKey (based on some
+   * hashing function) which can be used to index this.serverMap
+   *
+   * @param  {string} key
+   * * @returns {string}
```
Comment (blackmad) [inline src/memjs/memjs.ts]:
```suggestion
   * @returns {string}
```

File: src/memjs/memjs.ts [hunk 8]
```diff
@@ -0,0 +1,1777 @@
...
+    return new Client(servers, options as any);
+  }
+
+  /**
+   * Given a serverKey fromlookupKeyToServerKey, return the corresponding Server instance
+   *
+   * @param  {string} serverKey
+   * @returns {Server}
+   */
+  serverKeyToServer(serverKey: string) {
```
Comment (blackmad) [inline src/memjs/memjs.ts]:
```suggestion
  serverKeyToServer(serverKey: string): Server {
```

File: src/memjs/memjs.ts [hunk 9]
```diff
@@ -0,0 +1,1777 @@
...
+   * @param callback
+   */
+  get(key: string): Promise<{ value: Value | null; flags: Extras | null }>;
+  get(
+    key: string,
+    callback: (
+      error: Error | null,
+      value: Value | null,
+      flags: Extras | null
+    ) => void
```
Comment (blackmad) [inline src/memjs/memjs.ts:249]: surprised you didn't template this as Callback<Value>

File: src/memjs/memjs.ts [hunk 10]
```diff
@@ -0,0 +1,1777 @@
...
+  ): Promise<{ value: Value | null; flags: Extras | null }> | void {
+    var self = this;
+    if (callback === undefined) {
+      return promisify(function (callback) {
+        self.get(key, function (err, value, flags) {
+          callback(err, { value: value, flags: flags });
+        });
+      });
+    }
```
Comment (blackmad) [inline src/memjs/memjs.ts]: not for now,  but would be tempted to rip out all the callback code in this at some point to continue simplifying it

File: src/memjs/memjs.ts [hunk 11]
```diff
@@ -0,0 +1,1777 @@
...
+  ): void;
+  get(
+    key: string,
+    callback?: (
+      error: Error | null,
+      value: Value | null,
+      flags: Extras | null
+    ) => void
+  ): Promise<{ value: Value | null; flags: Extras | null }> | void {
+    var self = this;
```
Comment (blackmad) [inline src/memjs/memjs.ts]: could kill all the selfs

File: src/memjs/memjs.ts [hunk 12]
```diff
@@ -0,0 +1,1777 @@
...
+      if (err) {
+        if (callback) {
+          callback(err, null, null);
+        }
+        return;
+      }
+      switch (response!.header.status) {
+        case constants.RESPONSE_STATUS_SUCCCESS:
+          if (callback) {
+            var deserialized = self.serializer.deserialize(
```
Comment (blackmad) [inline src/memjs/memjs.ts]: surprised you didn't do a bulk find and replace for var -> const, and let vscode find all the ones that need to be lets

File: src/memjs/memjs.ts [hunk 13]
```diff
@@ -0,0 +1,1777 @@
...
+          }
+      }
+    });
+  }
+
+  /** Build a pipelined get multi request by sending one GETKQ for each key (quiet, meaning it won't respond if the value is missing) followed by a no-op to force a response (and to give us a sentinel response that the pipeline is done)
+   *
+   * cf https://github.com/couchbase/memcached/blob/master/docs/BinaryProtocol.md#0x0d-getkq-get-with-key-quietly
+   */
+  _buildGetMultiRequest(keys: string[]) {
```
Comment (blackmad) [inline src/memjs/memjs.ts:307]:
```suggestion
  _buildGetMultiRequest(keys: string[]): Buffer {
```

Comment (blackmad) [reply src/memjs/memjs.ts:307]: sorry I really like explicit return types to ensure readability and lack of bizarre bugs in the future

**Review by blackmad — APPROVED** [review]
becoming clear to me this is a partial rewrite, is fine, can continue cleaning in the future if we really want

File: src/memjs/server.ts [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+    this.writeSASL(buf);
+  }
+
+  appendToBuffer(dataBuf: Buffer) {
+    var old = this.responseBuffer;
+    this.responseBuffer = Buffer.alloc(old.length + dataBuf.length);
+    old.copy(this.responseBuffer, 0);
+    dataBuf.copy(this.responseBuffer, old.length);
+    return this.responseBuffer;
+  }
```
Comment (blackmad) [inline src/memjs/server.ts:152]:
```suggestion
  }
  
```

**Comment by blackmad** [comment]
I did all the let/const stuff, killed some old codepaths, and TSified the tests (all passing)

File: src/memjs/memjs.ts (hunk 4 above)
Comment (justjake) [reply src/memjs/memjs.ts:44]: I tried it out but native promisify wants to wrap the whole function. It doesn't quite match so I decided not to do it.

File: src/memjs/memjs.ts (hunk 5 above)
Comment (justjake) [reply src/memjs/memjs.ts:63]: This is fucky. Will comment

File: src/memjs/memjs.ts (hunk 13 above)
Comment (blackmad) [reply src/memjs/memjs.ts:307]: there is an eslint rule for this we could turn on

**Pull request merged**